scripts/gpu-box-setup/roles/common/tasks/main.yml

```diff
@@ -0,0 +1,208 @@
+---
+# System Setup and Hardening
+
+- name: Run setup to gather facts
+  setup: 
+    gather_subset: all
+
+# Sudo configuration is now handled in the pre-task play
+
+- name: Update package cache (RHEL family)
+  package:
+    update_cache: yes
+  when: ansible_os_family == 'RedHat'
+  
+- name: Upgrade all packages
+  package:
+    name: "*"
+    state: latest
+  when: system_upgrade | bool
+
+- name: Install essential packages
+  package:
+    name: "{{ essential_packages }}"
+    state: present
+
+# Python setup
+- name: Add deadsnakes PPA for Ubuntu
+  apt_repository:
+    repo: ppa:deadsnakes/ppa
+    state: present
+    update_cache: no  # We'll update cache separately to avoid timeout issues
+  register: deadsnakes_add
+  retries: 3
+  delay: 10
+  until: deadsnakes_add is succeeded
+  when: ansible_distribution == 'Ubuntu'
+  
+- name: Update apt cache (with increased timeout)
+  apt:
+    update_cache: yes
+    cache_valid_time: 86400  # Consider cache valid for 1 day
+  register: apt_update
+  retries: 3
+  delay: 10
+  until: apt_update is succeeded
+  when: ansible_os_family == 'Debian'
+
+- name: Install specified Python version
+  apt:
+    name:
+      - "python{{ python_version }}"
+      - "python{{ python_version }}-dev"
+      - "python{{ python_version }}-venv"
+    state: present
+  register: python_install
+  retries: 3
+  delay: 10
+  until: python_install is succeeded
+  when: ansible_os_family == 'Debian'
+
+- name: Install specified Python version on RHEL
+  dnf:
+    name:
+      - "python{{ python_version | regex_replace('\\.','')}}"
+      - "python{{ python_version | regex_replace('\\.','') }}-devel"
+    state: present
+  register: python_rhel_install
+  retries: 3
+  delay: 10
+  until: python_rhel_install is succeeded
+  when: ansible_os_family == 'RedHat'
```

Comment:
Also this is why `uv` is so great. It is managing both packages and python versions.

Reply:
point taken :)